cwlVersion: v1.1
class: CommandLineTool

# Extensions
$namespaces:
  s: https://schema.org/
  ilmn-tes: https://platform.illumina.com/rdf/ica/
$schemas:
  - https://schema.org/version/latest/schemaorg-current-http.rdf


# Metadata
s:author:
  class: s:Person
  s:name: Alexis Lucattini
  s:email: Alexis.Lucattini@umccr.org
  s:identifier: https://orcid.org/0000-0001-9754-647X


# ID/Docs
id: arriba-drawing--2.5.0
label: arriba-drawing v(2.5.0)
doc: |
  Documentation for arriba-drawing v2.5.0


# ILMN V1 Resources Guide: https://illumina.gitbook.io/ica-v1/analysis/a-taskexecution#type-and-size
# ILMN V2 Resources Guide: https://help.ica.illumina.com/project/p-flow/f-pipelines#compute-types
hints:
  ResourceRequirement:
    ilmn-tes:resources/tier: standard/economy
    ilmn-tes:resources/type: standard/standardHiCpu/standardHiMem/standardHiIo/fpga
    ilmn-tes:resources/size: large
    coresMin: 8
    ramMin: 28000
  DockerRequirement:
    dockerPull: public.ecr.aws/biocontainers/arriba:2.5.0--h87b9561_1


requirements:
  SchemaDefRequirement:
    types:
      - $import: ../../../schemas/bam-input/1.0.0/bam-input__1.0.0.yaml#bam-input
      - $import: ../../../schemas/cram-input/1.0.0/cram-input__1.0.0.yaml#cram-input
  InlineJavascriptRequirement: {}


baseCommand: [ "draw_fusions.R" ]


inputs:
  # Input files
  alignment_data:
    label: alignment data
    doc: |
      File in SAM/BAM/CRAM format with main alignments as generated by STAR
    type:
      - ../../../schemas/bam-input/1.0.0/bam-input__1.0.0.yaml#bam-input
      - ../../../schemas/cram-input/1.0.0/cram-input__1.0.0.yaml#cram-input
    inputBinding:
      prefix: --alignments=
      separate: false
      valueFrom: |
        ${
          if (self.bam_input) {
            return self.bam_input.path
          } else {
            return self.cram_input.path
          }
        }
  fusions_tsv:
    label: fusions
    doc: |
      TSV File containing fusion predictions from Arriba
    type: File
    inputBinding:
      prefix: --fusions=
      separate: false

  # References
  annotation_gtf:
    label: annotation
    doc: |
      Gene annotation in GTF format
    type: File
    inputBinding:
      prefix: --annotation=
      separate: false

  # Metadata
  output_pdf_filename:
    label: output_pdf_filename
    doc: |
      Output file in PDF format containing the visualizations of the gene fusions
    type: string?
    default: fusions.pdf
    inputBinding:
      prefix: --output=
      separate: false
  sample_name:
    label: sample name
    doc: |
      Sample name
    type: string?
    inputBinding:
      prefix: --sampleName=
      separate: false

  # Reference inputs
  cytobands_tsv:
    label: cytobands tsv
    doc: |
      Coordinates of the Giemsa staining bands. This information is used to draw ideograms
    type: File
    inputBinding:
      prefix: --cytobands=
      separate: false
  protein_domains_gff3:
    label: protein domains
    doc: |
      GFF3 file containing the genomic coordinates of protein domains
    type: File
    inputBinding:
      prefix: --proteinDomains=
      separate: false
  # Optional parameters
  squish_introns:
    label: squish introns
    doc: |
      Squish introns. Default true
    type: boolean?
    inputBinding:
      prefix: --squishIntrons=
      separate: false
  print_exon_labels:
    label: print exon labels
    doc: |
      Print exon labels. Default true
    type: boolean?
    inputBinding:
      prefix: --printExonLabels=
      separate: false
  render_3d_effect:
    label: render 3d effect
    doc: |
      Render 3d effect. Default true
    type: boolean?
    inputBinding:
      prefix: --render3dEffect=
      separate: false
  pdf_width:
    label: pdf width
    doc: |
      PDF width. Default 11.692
    type: int?
    inputBinding:
      prefix: --pdfWidth=
      separate: false
  color_1:
    label: color 1
    doc: |
      color 1. Default #e5a5a5
    type: string?
    inputBinding:
      prefix: --color1=
      separate: false
  color_2:
    label: color 2
    doc: |
      color 2. Default #a7c4e5
    type: string?
    inputBinding:
      prefix: --color2=
      separate: false
  merge_domains_overlapping_by:
    label: merge domains overlapping
    doc: |
      Merge domains overlapping. Default 0.9
    type: int?
    inputBinding:
      prefix: --mergeDomainsOverlappingBy=
      separate: false
  optimize_domain_colors:
    label: optimize domain colors
    doc: |
      Optimize domain colors. Default false.
    type: boolean?
    inputBinding:
      prefix: --optimizeDomainColors=
      separate: false
  font_size:
    label: font size
    doc: |
      Font size. Default 1.
    type: int?
    inputBinding:
      prefix: --fontSize=
      separate: false
  font_family:
    label: font family
    doc: |
      Font family. Default Helvetica.
    type: string?
    inputBinding:
      prefix: --fontFamily=
      separate: false
  transcript_selection:
    label: transcript selection
    doc: |
      Transcript selection. Default coverage
    type: string?
    inputBinding:
      prefix: --transcriptSelection=
      separate: false
  fixed_scale:
    label: fixed scale
    doc: |
      Fixed scale. Default 0.
    type: int?
    inputBinding:
      prefix: --fixedScale=
      separate: false
  coverage_range:
    label: coverage range
    doc: |
      Coverage range. Default 0.
    type: int?
    inputBinding:
      prefix: --coverageRange=
      separate: false
  plot_panels:
    label: plot panels
    doc: |
      Plot panels. Default 1.
    type: int?
    inputBinding:
      prefix: --plotPanels=
      separate: false


outputs:
  output_pdf:
    label: output PDF
    doc: |
      Output pdf file with fusion drawing
    type: File
    outputBinding:
      glob: "$(inputs.output_pdf_filename)"


successCodes:
  - 0