
arriba-fusion-calling 2.3.0 tool
================================

## Table of Contents
  
- [Overview](#arriba-fusion-calling-v230-overview)  
- [Links](#related-links)  
- [Inputs](#arriba-fusion-calling-v230-inputs)  
- [Outputs](#arriba-fusion-calling-v230-outputs)  
- [ICA](#ica)  


## arriba-fusion-calling v(2.3.0) Overview



  
> ID: arriba-fusion-calling--2.3.0  
> md5sum: b16400232a2fe17ad7d1d686ee985211

### arriba-fusion-calling v(2.3.0) documentation
  
Arriba is a command-line tool for the detection of gene fusions from RNA-Seq data  
https://github.com/suhrig/arriba.

### Categories
  


## Related Links
  
- [CWL File Path](../../../../../../tools/arriba-fusion-calling/2.3.0/arriba-fusion-calling__2.3.0.cwl)  


### Used By
  
- [dragen-transcriptome-pipeline 3.9.3](../../../workflows/dragen-transcriptome-pipeline/3.9.3/dragen-transcriptome-pipeline__3.9.3.md)  

  


## arriba-fusion-calling v(2.3.0) Inputs

### annotation



  
> ID: annotation
  
**Optional:** `False`  
**Type:** `File`  
**Docs:**  
GTF file with gene annotation. The file may be gzip-compressed


### bam file



  
> ID: bam_file
  
**Optional:** `False`  
**Type:** `File`  
**Docs:**  
File in SAM/BAM/CRAM format with main alignments as generated by STAR


### bam without chimeric alignments



  
> ID: bam_without_chimeric_alignments
  
**Optional:** `True`  
**Type:** `File`  
**Docs:**  
File in SAM/BAM/CRAM format with chimeric alignments as generated by STAR
(Chimeric.out.sam). This parameter is only required, if STAR was run with the
parameter '--chimOutType SeparateSAMold'. When STAR was run with the parameter
'--chimOutType WithinBAM', it suffices to pass the parameter -x to Arriba and -c
can be omitted.


### blacklist



  
> ID: blacklist
  
**Optional:** `False`  
**Type:** `File`  
**Docs:**  
File containing blacklisted ranges


### contigs



  
> ID: contigs
  
**Optional:** `True`  
**Type:** `string`  
**Docs:**  
Comma-/space-separated list of interesting contigs. Fusions between genes on other contigs are ignored


### coordinates protein domains



  
> ID: coordinates_protein_domains
  
**Optional:** `True`  
**Type:** `File`  
**Docs:**  
File in GFF3 format containing coordinates of the protein domains of genes. The
protein domains retained in a fusion are listed in the column
'retained_protein_domains'. The file may be gzip-compressed.


### coordinates structural variants



  
> ID: coordinates_structural_variants
  
**Optional:** `True`  
**Type:** `File`  
**Docs:**  
Tab-separated file with coordinates of structural variants found using
whole-genome sequencing data. These coordinates serve to increase sensitivity
towards weakly expressed fusions and to eliminate fusions with low evidence.


### covered fraction



  
> ID: covered_fraction
  
**Optional:** `True`  
**Type:** `int`  
**Docs:**  
Ignore virally associated events if the virus is not fully
expressed, i.e., less than the given fraction of the viral contig is
transcribed. Default: 0.050000


### discarded fusion



  
> ID: discarded_fusions_filename
  
**Optional:** `True`  
**Type:** `string`  
**Docs:**  
Output file with fusions that were discarded due to filtering


### exonic fraction



  
> ID: exonic_fraction
  
**Optional:** `True`  
**Type:** `int`  
**Docs:**  
The breakpoints of false-positive predictions of intragenic events
are often both in exons. True predictions are more likely to have at
least one breakpoint in an intron, because introns are larger. If the
fraction of exonic sequence between two breakpoints is smaller than
the given fraction, the 'intragenic_exonic' filter discards the
event. Default: 0.330000


### filters



  
> ID: filters
  
**Optional:** `True`  
**Type:** `string`  
**Docs:**  
Comma-/space-separated list of filters to disable. By default all filters are
enabled. Valid values: top_expressed_viral_contigs, viral_contigs,
low_coverage_viral_contigs, uninteresting_contigs,
no_genomic_support, short_anchor, select_best, many_spliced, long_gap,
marginal_read_through, merge_adjacent, hairpin, small_insert_size,
same_gene, mismappers, spliced, genomic_support, read_through,
no_coverage, mismatches, homopolymer, low_entropy, multimappers,
inconsistently_clipped, duplicates, homologs, blacklist,
relative_support, min_support, end_to_end, known_fusions,
non_coding_neighbors, isoforms, intronic, in_vitro, intragenic_exonic,
internal_tandem_duplication


### fragment length



  
> ID: fragment_length
  
**Optional:** `True`  
**Type:** `int`  
**Docs:**  
When paired-end data is given, the fragment length is estimated
automatically and this parameter has no effect. But when single-end
data is given, the mean fragment length should be specified to
effectively filter fusions that arise from hairpin structures.
Default: 200


### fusion



  
> ID: fusions_filename
  
**Optional:** `True`  
**Type:** `string`  
**Docs:**  
Output file with fusions that have passed all filters


### genomic breakpoint



  
> ID: genomic_breakpoint
  
**Optional:** `True`  
**Type:** `int`  
**Docs:**  
When a file with genomic breakpoints obtained via whole-genome sequencing is supplied via the -d
parameter, this parameter determines how far a genomic breakpoint may be away from a
transcriptomic breakpoint to consider it as a related event. For events inside genes, the
distance is added to the end of the gene; for intergenic events, the distance threshold is
applied as is. Default: 100000


### GTF features



  
> ID: gtf_features
  
**Optional:** `True`  
**Type:** `string`  
**Docs:**  
Comma-/space-separated list of names of GTF features.
Default: gene_name=gene_name|gene_id gene_id=gene_id
transcript_id=transcript_id feature_exon=exon feature_CDS=CDS


### homopolymer length



  
> ID: homopolymer_length
  
**Optional:** `True`  
**Type:** `int`  
**Docs:**  
The 'homopolymer' filter removes breakpoints adjacent to
homopolymers of the given length or more. Default: 6


### known fusions



  
> ID: known_fusions
  
**Optional:** `True`  
**Type:** `File`  
**Docs:**  
File containing known/recurrent fusions. Some cancer entities are often
characterized by fusions between the same pair of genes. In order to boost
sensitivity, a list of known fusions can be supplied using this parameter. The list
must contain two columns with the names of the fused genes, separated by tabs.


### many spliced events



  
> ID: many_spliced_events
  
**Optional:** `True`  
**Type:** `int`  
**Docs:**  
The 'many_spliced' filter recovers fusions between genes that
have at least this many spliced breakpoints. Default: 4


### max e value



  
> ID: max_e_value
  
**Optional:** `True`  
**Type:** `int`  
**Docs:**  
Arriba estimates the number of fusions with a given number of supporting
reads which one would expect to see by random chance. If the expected number
of fusions (e-value) is higher than this threshold, the fusion is
discarded by the 'relative_support' filter. Note: Increasing this
threshold can dramatically increase the number of false positives and may
increase the runtime of resource-intensive steps. Fractional values are
possible. Default: 0.300000


### max homolog identity



  
> ID: max_homolog_identity
  
**Optional:** `True`  
**Type:** `int`  
**Docs:**  
Genes with more than the given fraction of sequence identity are
considered homologs and removed by the 'homologs' filter.
Default: 0.300000


### max itd length



  
> ID: max_itd_length
  
**Optional:** `True`  
**Type:** `int`  
**Docs:**  
Maximum length of internal tandem duplications. Note: Increasing
this value beyond the default can impair performance and lead to many
false positives. Default: 100


### max kmer content



  
> ID: max_kmer_content
  
**Optional:** `True`  
**Type:** `int`  
**Docs:**  
The 'low_entropy' filter removes reads with repetitive 3-mers. If
the 3-mers make up more than the given fraction of the sequence, then
the read is discarded. Default: 0.600000


### max mismappers



  
> ID: max_mismappers
  
**Optional:** `True`  
**Type:** `int`  
**Docs:**  
When more than this fraction of supporting reads turns out to be
mismappers, the 'mismappers' filter discards the fusion. Default:
0.800000


### max mismatch pvalue



  
> ID: max_mismatch_pvalue
  
**Optional:** `True`  
**Type:** `int`  
**Docs:**  
The 'mismatches' filter uses a binomial model to calculate a
p-value for observing a given number of mismatches in a read. If
the number of mismatches is too high, the read is discarded.
Default: 0.010000


### max reads



  
> ID: max_reads
  
**Optional:** `True`  
**Type:** `int`  
**Docs:**  
Subsample fusions with more than the given number of supporting reads. This
improves performance without compromising sensitivity, as long as the
threshold is high. Counting of supporting reads beyond the threshold is
inaccurate, obviously. Default: 300


### min anchor length



  
> ID: min_anchor_length
  
**Optional:** `True`  
**Type:** `int`  
**Docs:**  
Alignment artifacts are often characterized by split reads coming
from only one gene and no discordant mates. Moreover, the split
reads only align to a short stretch in one of the genes. The
'short_anchor' filter removes these fusions. This parameter sets
the threshold in bp for what the filter considers short. Default: 23


### min itd allele fraction



  
> ID: min_itd_allele_fraction
  
**Optional:** `True`  
**Type:** `int`  
**Docs:**  
Required fraction of supporting reads to report an internal
tandem duplication. Default: 0.070000


### min itd supporting reads



  
> ID: min_itd_supporting_reads
  
**Optional:** `True`  
**Type:** `int`  
**Docs:**  
Required absolute number of supporting reads to report an
internal tandem duplication. Default: 10


### min supporting reads



  
> ID: min_supporting_reads
  
**Optional:** `True`  
**Type:** `int`  
**Docs:**  
The 'min_support' filter discards all fusions with fewer than
this many supporting reads (split reads and discordant mates
combined). Default: 2


### quantile



  
> ID: quantile
  
**Optional:** `True`  
**Type:** `int`  
**Docs:**  
Highly expressed genes are prone to produce artifacts during library
preparation. Genes with an expression above the given quantile are eligible
for filtering by the 'in_vitro' filter. Default: 0.998000


### read through distance



  
> ID: read_through_distance
  
**Optional:** `True`  
**Type:** `int`  
**Docs:**  
The 'read_through' filter removes read-through fusions
where the breakpoints are less than the given distance away
from each other. Default: 10000


### reference



  
> ID: reference
  
**Optional:** `False`  
**Type:** `File`  
**Docs:**  
FastA file with genome sequence (assembly). The file may be gzip-compressed


### strandedness



  
> ID: strandedness
  
**Optional:** `True`  
**Type:** `string`  
**Docs:**  
Whether a strand-specific protocol was used for library preparation,
and if so, the type of strandedness (auto/yes/no/reverse). When
unstranded data is processed, the strand can sometimes be inferred from
splice-patterns. But in unclear situations, stranded data helps
resolve ambiguities. Default: auto


### tags



  
> ID: tags
  
**Optional:** `True`  
**Type:** `File`  
**Docs:**  
Tab-separated file containing fusions to annotate with tags in the 'tags' column.
The first two columns specify the genes; the third column specifies the tag. The
file may be gzip-compressed.


### top N



  
> ID: top_n
  
**Optional:** `True`  
**Type:** `int`  
**Docs:**  
Only report viral integration sites of the top N most highly expressed viral
contigs. Default: 5


### viral contigs



  
> ID: viral_contigs
  
**Optional:** `True`  
**Type:** `string`  
**Docs:**  
Comma-/space-separated list of viral contigs. Asterisks (*) are treated as
wild-cards.
Default: AC_* NC_*

  


## arriba-fusion-calling v(2.3.0) Outputs

### discarded fusions



  
> ID: arriba-fusion-calling--2.3.0/discarded_fusions  

  
**Optional:** `False`  
**Output Type:** `File`  
**Docs:**  
file with discarded fusions
  


### fusions



  
> ID: arriba-fusion-calling--2.3.0/fusions  

  
**Optional:** `False`  
**Output Type:** `File`  
**Docs:**  
file with output fusions
  

  


## ICA

### ToC
  
- [development_workflows](#project-development_workflows)  


### Project: development_workflows


> wfl id: wfl.b9a507a2bfc549a6ba82db33d7d64d9c  

  
**workflow name:** arriba-fusion-calling_dev-wf  
**wfl version name:** 2.3.0  

  

